### ISSUE TITLE
Info request - best practice for config custom datasets?

### ISSUE THREAD

**Issue by lessw2020** [rwightman/efficientdet-pytorch#3]
Hi @rwightman  - super excited that you have made this! 
Can you describe best method for training on custom datasets? (I can hack it up but prefer if you have have advice up front).
Thanks!

**Comment by lessw2020** [comment]
I see the config has a num_classes setting, so setting that for now but also see the #Fixme re weight inits for training.

**Comment by rwightman** [comment]
@lessw2020 
Num classes is just a very small part of it. The weight init doesn't actually matter if you fine-tune from the pretrained weights. You just need the weight init of the last class head layer sorted.

There is no training code right now, it's quite a bit more complicated than classification so you'll need to get familiar with retinanet style bbox anchors if you want to get training running. You could likely lift something from other PyTorch retinanet or efficientdet code.

I think I've translated most of the necessary code for replication of the official but it's not at all tested, and no base training code written yet, but easy enough to lift that from somewhere. The plan was to create another 'Bench' class, a `DetBenchTrain`. It will use the AnchorLabeler with passed in annotation targets from the train dataset to generate bbox regression targets.

See original (I was not planning on doing that in the dataset/loader like they do)
https://github.com/rwightman/efficientdet-pytorch/blob/master/effdet/anchors.py#L370
https://github.com/google/automl/blob/269710783e4a7da55b0e8dc68a2ca9dea7fa2baf/efficientdet/dataloader.py#L239

The dataset is spitting out targets but I haven't verified they are 100% correct and in the right yxyx order, scaled properly, etc.

Also, the losses between the network output and the generated anchor targets need to be setup, but that should be pretty easy to translate from the TF original.

I also need to add some augmentation transforms, fairly simple without autoaugment, just need a scale and x/y offset transform with bbox scale/clip. 

I plan to get to all of this at some point, but looks like I'll have some other priorities coming up short term....

**Comment by lessw2020** [comment]
Hi @rwightman  - thanks very much for the update and info!  I've been working with the signatrix and TF D0 already, and going through your code now (your code is super clean btw, pleasure to read through it.)
But yes I did see that it's not hooked up for training yet after I started working through it.
I'd be interested to help push your goal of the DetBenchTrain forward.  I'll see if I can patch together some quick code leveraging from other impls to get a basic training framework and then can PR it if it's working well.

**Comment by rwightman** [comment]
@lessw2020 sure, contribs welcome, interested to see what you come up with

**Comment by Dicko87** [comment]
Hi guys, I have a custom dataset that I would like to use with Efficientdet and know I need to fine tune the network. Has anyone trained with a custom dataset  yet? If so, could you point me in the right direction of how to go about it please. Thanks very much

**Comment by tazu786** [comment]
> 
> 
> Hi guys, I have a custom dataset that I would like to use with Efficientdet and know I need to fine tune the network. Has anyone trained with a custom dataset yet? If so, could you point me in the right direction of how to go about it please. Thanks very much

I would be very interested too!

**Comment by jinfagang** [comment]
@Dicko87 No one, I tried train on 2 custom dataset, both result like this:

![image](https://user-images.githubusercontent.com/21303438/95649056-36b1c600-0b0e-11eb-8758-2959bcbd5551.png)

The issue could be:
 a. Anchor is all wrong, training code is messed up and not correclty at all;
b. I miss configred something

**Comment by Ekta246** [comment]
> @Dicko87 No one, I tried train on 2 custom dataset, both result like this:
> 
> ![image](https://user-images.githubusercontent.com/21303438/95649056-36b1c600-0b0e-11eb-8758-2959bcbd5551.png)
> 
> The issue could be:
> a. Anchor is all wrong, training code is messed up and not correclty at all;
> b. I miss configred something

Looks like you managed to run the evaluation metrics such as mAP for your custom dataset? Could you please tell me if you had restructured your custom dataset in the Coco annotations and hence use coco_api format or you managed to write a custom dataloader for your custom dataset and yet manage to use the coco_api.

Looking forward to your response.
Thanks

**Comment by Ekta246** [comment]
> @Dicko87 No one, I tried train on 2 custom dataset, both result like this:
> 
> ![image](https://user-images.githubusercontent.com/21303438/95649056-36b1c600-0b0e-11eb-8758-2959bcbd5551.png)
> 
> The issue could be:
> a. Anchor is all wrong, training code is messed up and not correclty at all;
> b. I miss configred something

Also, did you try training the model with bounding box annotations in the format{ymin, xmin, ymax, xmax}?

**Comment by rwightman** [comment]
Fine-tuning w/ changeable class size and proper re-init of the head is in the master now. I've compared fine tuning from coco weights vs from scratch w/ the two new added datasets (Pascal VOC, OpenImages) and Pascal trained to end, works well. OI is huge so I just started it but was off to a good start. It should be easy enough to cut and paste aspects of the 3 different dataset annotation types COCO, VOC, OpenImages to support other custom datasets, or just use the parsers directly w/ new config dataclasses if they use similar JSON, XML, CSV formats.
